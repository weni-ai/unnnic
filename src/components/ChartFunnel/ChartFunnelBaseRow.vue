<template>
  <svg
    class="unnnic-chart-funnel-base-rows"
    :viewBox="viewBox"
    fill="none"
    xmlns="http://www.w3.org/2000/svg"
  >
    <g
      v-for="(row, index) of rows"
      :key="row.title + index"
      class="unnnic-chart-funnel-base-rows__row"
    >
      <path
        class="row__path"
        :d="row.pathD"
      />
      <text
        x="50%"
        :y="row.titleY"
        class="row__title"
      >
        {{ row.title }}
      </text>
      <text
        x="50%"
        :y="row.descriptionY"
        class="row__description"
      >
        {{ row.description }}
      </text>
    </g>
  </svg>
</template>

<script>
export default {
  name: 'UnnnicChartFunnelThreeRows',

  props: {
    rows: {
      type: Array,
      required: true,
    },

    viewBox: String,
  },
};
</script>

<style lang="scss" scoped>
@import '../../assets/scss/unnnic.scss';

.unnnic-chart-funnel-base-rows {
  width: 100%;
  height: 100%;
  margin: 0;
  padding: 0;

  &__row {
    .row__title,
    .row__description {
      fill: $unnnic-color-neutral-white;
      font-family: $unnnic-font-family-secondary;
      text-anchor: middle;
    }

    .row__title {
      font-size: $unnnic-font-size-body-lg;
      font-weight: $unnnic-font-weight-bold;
    }

    .row__description {
      font-size: $unnnic-font-size-body-md;
    }
  }
}
</style>
