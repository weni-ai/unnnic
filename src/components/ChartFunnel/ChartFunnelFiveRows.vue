<template>
  <ChartFunnelBaseRow
    class="unnnic-chart-funnel-five-rows"
    :rows="rows"
    viewBox="0 0 361 220"
  />
</template>

<script>
import ChartFunnelBaseRow from './ChartFunnelBaseRow.vue';
export default {
  name: 'UnnnicChartFunnelFiveRows',

  components: {
    ChartFunnelBaseRow,
  },

  props: {
    data: {
      type: Array,
      required: true,
    },
  },

  computed: {
    rows() {
      const { data } = this;
      return [
        {
          pathD:
            'M352.506 0H8.42251C1.93921 0 -1.85142 7.30766 1.88266 12.6076L24 44H336.75L359.028 12.6324C362.79 7.33528 359.003 0 352.506 0Z',
          titleX: '50%',
          titleY: '20',
          descriptionX: '50%',
          descriptionY: '38',
          title: data[0].title,
          description: data[0].description,
        },
        {
          pathD: 'M337 44H24L56.4654 88H304.535L337 44Z',
          titleX: '50%',
          titleY: '62',
          descriptionX: '50%',
          descriptionY: '80',
          title: data[1].title,
          description: data[1].description,
        },
        {
          pathD: 'M304.5 88H56.5L88.9564 132H273.708L304.5 88Z',
          titleX: '50%',
          titleY: '106',
          descriptionX: '50%',
          descriptionY: '124',
          title: data[2].title,
          description: data[2].description,
        },
        {
          pathD: 'M274 132H89L121 176H242L274 132Z',
          titleX: '50%',
          titleY: '150',
          descriptionX: '50%',
          descriptionY: '168',
          title: data[3].title,
          description: data[3].description,
        },
        {
          pathD:
            'M242 176H120.957L130.481 213.947C131.374 217.505 134.573 220 138.241 220H223.89C227.498 220 230.659 217.585 231.608 214.105L242 176Z',
          titleX: '50%',
          titleY: '194',
          descriptionX: '50%',
          descriptionY: '212',
          title: data[4].title,
          description: data[4].description,
        },
      ];
    },
  },
};
</script>

<style lang="scss">
@import '../../assets/scss/unnnic.scss';

.unnnic-chart-funnel-five-rows {
  [class$='row']:nth-child(1) {
    fill: $unnnic-color-weni-950;
  }
  [class$='row']:nth-child(2) {
    fill: $unnnic-color-weni-900;
  }
  [class$='row']:nth-child(3) {
    fill: $unnnic-color-weni-800;
  }
  [class$='row']:nth-child(4) {
    fill: $unnnic-color-weni-700;
  }
  [class$='row']:nth-child(5) {
    fill: $unnnic-color-weni-600;
  }
}
</style>
