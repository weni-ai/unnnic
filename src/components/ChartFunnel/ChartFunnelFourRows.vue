<template>
  <ChartFunnelBaseRow
    class="unnnic-chart-funnel-four-rows"
    :rows="rows"
    viewBox="0 0 365 224"
  />
</template>

<script>
import ChartFunnelBaseRow from './ChartFunnelBaseRow.vue';
export default {
  name: 'UnnnicChartFunnelFourRows',

  components: {
    ChartFunnelBaseRow,
  },

  props: {
    data: {
      type: Array,
      required: true,
    },
  },

  computed: {
    rows() {
      const { data } = this;
      return [
        {
          pathD:
            'M356.374 0H8.57255C2.47804 0 -1.37827 6.54249 1.5734 11.8745L26 56H338.75L363.36 11.8984C366.336 6.56589 362.481 0 356.374 0Z',
          titleX: '50%',
          titleY: '23',
          descriptionX: '50%',
          descriptionY: '43',
          title: data[0].title,
          description: data[0].description,
        },
        {
          pathD: 'M339 56H26L58.4654 112H306.535L339 56Z',
          titleX: '50%',
          titleY: '78',
          descriptionX: '50%',
          descriptionY: '98',
          title: data[1].title,
          description: data[1].description,
        },
        {
          pathD: 'M306.5 112H58.5L90.9564 168H275.708L306.5 112Z',
          titleX: '50%',
          titleY: '136',
          descriptionX: '50%',
          descriptionY: '156',
          title: data[2].title,
          description: data[2].description,
        },
        {
          pathD:
            'M276 168H91L120.697 219.969C122.121 222.462 124.772 224 127.643 224H239.357C242.228 224 244.879 222.462 246.303 219.969L276 168Z',
          titleX: '50%',
          titleY: '190',
          descriptionX: '50%',
          descriptionY: '210',
          title: data[3].title,
          description: data[3].description,
        },
      ];
    },
  },
};
</script>

<style lang="scss">
@import '../../assets/scss/unnnic.scss';

.unnnic-chart-funnel-four-rows {
  [class$='row']:nth-child(1) {
    fill: $unnnic-color-weni-950;
  }
  [class$='row']:nth-child(2) {
    fill: $unnnic-color-weni-800;
  }
  [class$='row']:nth-child(3) {
    fill: $unnnic-color-weni-700;
  }
  [class$='row']:nth-child(4) {
    fill: $unnnic-color-weni-600;
  }
}
</style>
