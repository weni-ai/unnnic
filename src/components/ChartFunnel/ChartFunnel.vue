<template>
  <component
    :is="chartComponent"
    :data="chartData"
  />
</template>

<script>
import ChartFunnelThreeRows from './SvgFunnel/ChartFunnelThreeRows.vue';
import ChartFunnelFourRows from './SvgFunnel/ChartFunnelFourRows.vue';
import ChartFunnelFiveRows from './SvgFunnel/ChartFunnelFiveRows.vue';
import ChartDefaultFunnelThreeRows from './DefaultFunnel/ChartDefaultFunnelThreeRows.vue';
import ChartDefaultFunnelFourRows from './DefaultFunnel/ChartDefaultFunnelFourRows.vue';
import ChartDefaultFunnelFiveRows from './DefaultFunnel/ChartDefaultFunnelFiveRows.vue';

export default {
  name: 'UnnnicChartFunnel',

  components: {
    ChartFunnelThreeRows,
    ChartFunnelFourRows,
    ChartFunnelFiveRows,
    ChartDefaultFunnelFiveRows,
    ChartDefaultFunnelThreeRows,
    ChartDefaultFunnelFourRows,
  },

  props: {
    data: {
      type: Array,
      required: true,
    },
    type: {
      type: String,
      default: 'default',
    },
  },

  computed: {
    chartComponent() {
      const componentMap = {
        default: {
          3: ChartDefaultFunnelThreeRows,
          4: ChartDefaultFunnelFourRows,
          5: ChartDefaultFunnelFiveRows,
        },
        basic: {
          3: 'ChartFunnelThreeRows',
          4: 'ChartFunnelFourRows',
          5: 'ChartFunnelFiveRows',
        },
      };
      return componentMap[this.type][this.data.length] || null;
    },
    chartData() {
      const classIndex = ['w-60', 'w-50', 'w-40', 'w-30', 'w-20'];
      if (this.type === 'default')
        return this.data.map((e, index) => ({
          percentage: e.title,
          value: e.value,
          description: e.description,
          widthClass: classIndex[index],
          color: e.color,
        }));

      return this.data;
    },
  },
};
</script>
