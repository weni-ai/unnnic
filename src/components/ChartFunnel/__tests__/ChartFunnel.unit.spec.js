import { mount } from '@vue/test-utils';
import { describe, expect, test, beforeEach } from 'vitest';
import ChartFunnel from '@/components/ChartFunnel/ChartFunnel.vue';

describe('ChartFunnel.vue', () => {
  const mockThreeRowsData = [
    { title: '100%', value: 1000, description: 'First', color: '#123456' },
    { title: '66%', value: 660, description: 'Second', color: '#234567' },
    { title: '33%', value: 330, description: 'Third', color: '#345678' },
  ];

  const mockFourRowsData = [
    ...mockThreeRowsData,
    { title: '10%', value: 100, description: 'Fourth', color: '#456789' },
  ];

  const mockFiveRowsData = [
    ...mockFourRowsData,
    { title: '5%', value: 50, description: 'Fifth', color: '#567890' },
  ];

  let wrapper;

  beforeEach(() => {
    wrapper = mount(ChartFunnel, {
      props: {
        data: mockThreeRowsData,
        type: 'default',
      },
      global: {
        stubs: {
          ChartDefaultFunnelBase: true,
          ChartFunnelThreeRows: true,
          ChartFunnelFourRows: true,
          ChartFunnelFiveRows: true,
        },
      },
    });
  });

  test('renders correctly', () => {
    expect(wrapper.exists()).toBe(true);
  });

  test('validates required props', () => {
    const { data } = wrapper.vm.$options.props;
    expect(data.required).toBe(true);
    expect(data.type).toBe(Array);
  });

  test('validates default props', () => {
    const { type } = wrapper.vm.$options.props;
    expect(type.default).toBe('default');
    expect(type.type).toBe(String);
  });

  describe('chartComponent computed property', () => {
    test('returns correct component for default type with 3 rows', () => {
      expect(wrapper.vm.chartComponent.__name).toBe('ChartDefaultFunnelBase');
    });

    test('returns correct component for default type with 4 rows', async () => {
      await wrapper.setProps({ data: mockFourRowsData });
      expect(wrapper.vm.chartComponent.__name).toBe('ChartDefaultFunnelBase');
    });

    test('returns correct component for default type with 5 rows', async () => {
      await wrapper.setProps({ data: mockFiveRowsData });
      expect(wrapper.vm.chartComponent.__name).toBe('ChartDefaultFunnelBase');
    });

    test('returns correct component for basic type with 3 rows', async () => {
      await wrapper.setProps({ type: 'basic', data: mockThreeRowsData });
      expect(wrapper.vm.chartComponent).toBe('ChartFunnelThreeRows');
    });

    test('returns correct component for basic type with 4 rows', async () => {
      await wrapper.setProps({ type: 'basic', data: mockFourRowsData });
      expect(wrapper.vm.chartComponent).toBe('ChartFunnelFourRows');
    });

    test('returns correct component for basic type with 5 rows', async () => {
      await wrapper.setProps({ type: 'basic', data: mockFiveRowsData });
      expect(wrapper.vm.chartComponent).toBe('ChartFunnelFiveRows');
    });

    test('returns null for invalid row count', async () => {
      const invalidData = [
        { title: '100%', value: 100, description: 'Single' },
      ];
      await wrapper.setProps({ data: invalidData });
      expect(wrapper.vm.chartComponent).toBe(null);
    });
  });

  describe('chartData computed property', () => {
    test('transforms data correctly for default type', () => {
      const expectedData = [
        {
          percentage: '100%',
          value: 1000,
          description: 'First',
          widthClass: 'w-60',
          color: '#123456',
        },
        {
          percentage: '66%',
          value: 660,
          description: 'Second',
          widthClass: 'w-50',
          color: '#234567',
        },
        {
          percentage: '33%',
          value: 330,
          description: 'Third',
          widthClass: 'w-40',
          color: '#345678',
        },
      ];
      expect(wrapper.vm.chartData).toEqual(expectedData);
    });

    test('returns original data for basic type', async () => {
      await wrapper.setProps({ type: 'basic' });
      expect(wrapper.vm.chartData).toEqual(mockThreeRowsData);
    });

    test('applies correct width classes for different row counts in default type', async () => {
      await wrapper.setProps({ data: mockFiveRowsData });
      const widthClasses = wrapper.vm.chartData.map((item) => item.widthClass);
      expect(widthClasses).toEqual(['w-60', 'w-50', 'w-40', 'w-30', 'w-20']);
    });
  });

  test('renders the correct dynamic component based on computed chartComponent', async () => {
    let component = wrapper.findComponent({ name: 'ChartDefaultFunnelBase' });
    expect(component.exists()).toBe(true);

    await wrapper.setProps({ type: 'basic', data: mockThreeRowsData });
    component = wrapper.findComponent({ name: 'ChartFunnelThreeRows' });
    expect(component.exists()).toBe(true);

    await wrapper.setProps({ data: mockFourRowsData });
    component = wrapper.findComponent({ name: 'ChartFunnelFourRows' });
    expect(component.exists()).toBe(true);

    await wrapper.setProps({ data: mockFiveRowsData });
    component = wrapper.findComponent({ name: 'ChartFunnelFiveRows' });
    expect(component.exists()).toBe(true);
  });

  test('passes correct props to dynamic component', async () => {
    const component = wrapper.findComponent({ name: 'ChartDefaultFunnelBase' });
    expect(component.props('data')).toEqual(wrapper.vm.chartData);
  });

  test('matches snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });
});
