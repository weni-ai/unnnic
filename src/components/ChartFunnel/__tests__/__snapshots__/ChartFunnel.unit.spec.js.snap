// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ChartFunnel.vue > matches snapshot 1`] = `"<chart-default-funnel-base-stub data="[object Object],[object Object],[object Object]"></chart-default-funnel-base-stub>"`;
